pkg/proctree/tree.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		logger.Errorw("delete thread from tree", "err", err.Error())
+	}
+	tnode.DisconnectNode()
+	return nil
+}
+
+// addGeneralEventProcess generate a new process with information that could be received from any
+// event from the process
+func (tree *ProcessTree) addGeneralEventProcess(event *trace.Event) (*processNode, error) {
+	process, err := tree.newProcessNode(event.HostProcessID)
+	if err != nil {
+		return nil, err
+	}
+	process.setGenInfoFromEventProtected(event)
+	return process, nil
+}
+
+// addGeneralEventThread generate a new thread with information that could be received from any
+// event from the thread
+func (tree *ProcessTree) addGeneralEventThread(event *trace.Event) (*threadNode, error) {
+	p, err := tree.getProcess(event.HostProcessID)
+	if err != nil {
+		p, err = tree.addGeneralEventProcess(event)
+		if err != nil {
+			return nil, err
+		}
+	}
+	p.mutex.Lock()
+	thread, err := tree.newProcessThreadNode(p, event.HostThreadID)
+	exitTime := p.getExitTime()
+	p.mutex.Unlock()
+	if err != nil {
+		return nil, err
+	}
+
+	thread.setGenInfoFromEventProtected(event, exitTime)
+	return thread, nil
+}
+
+// generateParentProcess add a parent process to given process from tree if existing or creates
+// new node with the best effort info
+func (tree *ProcessTree) generateParentProcess(parentHostId int, parentNsId int, process *processNode) (*processNode, error) {
+	if parentNsId != 0 &&
+		process.getId() != parentHostId { // Prevent looped references
+		parentProcess, err := tree.getProcess(parentHostId)
+		if err != nil {
+			// TODO: Fix the race condition here between checking if exist and setting new one
+			parentProcess, err = tree.newProcessNode(parentHostId)
+			if err != nil {
+				return nil, err
+			}
+			parentProcess.setNsId(parentNsId)
+			if err != nil {
+				return nil, err
+			}
+		}
+		process.mutex.Lock()
+		process.connectParent(parentProcess)
+		process.mutex.Unlock()
+		parentProcess.mutex.Lock()
+		parentProcess.connectChild(process)
+		parentProcess.mutex.Unlock()
+	}
+	return process, nil
+}
+
+// newProcessNode create a new processNode and sign it in the tree
+func (tree *ProcessTree) newProcessNode(pid int) (*processNode, error) {
+	proc, err := newProcessNode(pid)
+	if err != nil {
+		return nil, err
+	}
+	tree.processes.Set(pid, proc)
+	tree.processesGC.Add(proc.GetUniqueId(), proc)
+	return proc, nil
+}
+
+// newThreadNode create a new threadNode and sign it in the tree
+func (tree *ProcessTree) newThreadNode(tid int) (*threadNode, error) {
+	thread, err := newThreadNode(tid)
+	if err != nil {
+		return nil, err
+	}
+	tree.threads.Set(tid, thread)
+	tree.threadsGC.Add(thread.GetUniqueId(), thread)
+	return thread, nil
+}
+
+// newProcessThreadNode add a new thread to a process, and also sign the thread in the tree.
+// It will return existing one if there is one with the same ID.
+func (tree *ProcessTree) newProcessThreadNode(process *processNode, tid int) (*threadNode, error) {
```

Comment:
The name is misleading - it seams like the thread will always be created, not only if it doesn't exist.
I think a better name would be `getOrCreateNewProcessThreadNode` or `createIfMissingProcessThreadNode`.